snapcraft/internal/lifecycle.py

```diff
@@ -243,6 +244,18 @@ def _run_step(self, step, part, part_names):
         common.env.extend(self.config.project_env())
 
         part = _replace_in_part(part)
+
+        # Add the annotated list of build packages
+        part_build_packages = part._part_properties['build-packages']
+        part.build_packages = []
+        for pkg in self.build_packages:
```

Comment:
it's adding `pkg` to the parts build-package list, which would match for `baz=3.4` but not for `bar=1.2` since it is not in the parts build-package list from the originial snapcraft.yaml.  Neither `bar` nor `bar=1.2` will match within the loop and thus won't (correctly) get added to the part's build-package list in the assets dict.